### DISCUSSION TITLE
Test hierarchy using local functions

### DISCUSSION THREAD

**Discussion by alataffective** [xunit/xunit#2194] [Ideas]
**Idea**
Allow tests to be defined using nested local functions.

**Minimal Example**
```
public class Tests
{
    void Arrange()
    {
        var sut = new Sut();
        
        void Act()
        {
            sut.Act();

            [Fact] void AssertTrue() => Assert.True(sut.State);
        }
    }
}
```

**Rationale**
When testing, I often find myself with a conceptual hierarchy of test cases. When doing so, I like to reuse setup code, and organise my tests so that it easy to see the hierarchy in the code.

A simple hierarchy may be as follows:
```
Given 1
    When 1.1
        Then 1.1.1
        Then 1.1.2
    And 1.2
        When 1.2.1
            Then 1.2.1.1
```

In order to reuse the setup logic in `Given 1`, and sticking to a single assertion per test method, I can implement this in xUnit as:

```
public class Given1
{
    Sut _sut;

    public Given1()
    {
        _sut = new Sut();
    }

    [Fact]
    void When11Then111()
    {
        _sut.When11();
        
        Assert.True(_sut.Then111);
    }

    [Fact]
    void When11Then112()
    {
        _sut.When11();
        
        Assert.True(_sut.Then112);
    }

    [Fact]
    void And12When121Then1211()
    {
        _sut.And12();

        _sut.When121();
        
        Assert.True(_sut.Then1211);
    }
}
```

However, in this flat form, this doesn't scale very well to even slightly more complex hierarchies, as the tree structure is lost.

Hence, despite the additional code, I often feel happier organising my tests using subclassing.

```
public class Given1
{
    Sut _sut;

    public Given1()
    {
        _sut = new Sut();
    }

    public class When11 : Given1
    {
        public When11() => _sut.When11();

        [Fact] void Then111() => Assert.True(_sut.Then111);
        [Fact] void Then112() => Assert.True(_sut.Then112);
    }

    public class And12 : Given1
    {
        public And12() => _sut.And12();

        public class When121 : And12
        {
            public When121() => _sut.When121();

            [Fact] void Then1211() => Assert.True(_sut.Then1211);
        }
    }
}
```

I find this an improvement over trying to represent a hierarchy using the flatter form, as it shares test setup code, organises the tests into a tree, makes names shorter, and allows the use of editor code folding features to collapse and expand tree nodes.

The downside of doing it this way is that there is more code needed to support this structural organisation. Additionally, when refactoring tests, it is necessary to remember to make sure the inner classes inherit from the outer class, otherwise bad things happen.

There may be other ways of doing something similar that is cleaner than this approach. I'd love to hear ideas. Going forward though, with the ability to declare attributes on local functions coming with C# 9.0, it occurs to me that xUnit could potentially support the following.

```
public class Given1
{
    Sut _sut;

    public Given1()
    {
        _sut = new Sut();
    }

    void When11()
    {
        _sut.When11();

        [Fact] void Then111() => Assert.True(_sut.Then111);
        [Fact] void Then112() => Assert.True(_sut.Then112);
    }

    void And12()
    {
        _sut.And12();

        void When121()
        {
            _sut.When121();

            [Fact] void Then1211() => Assert.True(_sut.Then1211);
        }
    }
}
```

Is this desirable? What would be needed to make this happen?

**Comment by bradwilson** [comment]
One problem I see is that you have implied dependency order here, which xUnit.net works very hard against (test case run order is randomized, explicitly to prevent this kind of dependency).

I think maybe a different testing framework that sits on top of xUnit.net (like [xUnit.net BDD Extensions](https://github.com/chadly/xunit-bdd)) might better suit your code style?

**Reply by alataffective** [reply]
Thanks for this Brad.

The order in which the tests are run is not important, only that the steps that occur before the execution of the `[Fact]` method are performed in the correct order for each test. Exactly as they are when using the inner subclass arrangement shown above.

I haven't tried xUnit.net BDD Extensions, although a quick look at the project doesn't seem to suggest a way of managing and sharing the tree of setups and actions. All of the examples shown are flat (unless I've missed something).

What I'm trying to do is structure the tree of setups and actions, in the way that is achieved using the inner subclass arrangement, but with less ceremony, noise and potential for errors. The randomization of test order is not affected.

**Reply by bradwilson** [reply]
I'm going to mark this as answered, because this isn't really a core feature for us (but rather I think a search for an alternate way to express tests which coincidentally reuses the xUnit.net infrastructure). Let me know if you have any further questions.

**Reply by alataffective** [reply]
Ok Brad, not a core feature I understand.

I suppose if I wanted to implement this myself, I'd need support from the test runner? I don't know enough about how xUnit tests are discovered and executed to be able to make any further guesses at this point. Can you make a suggestion regarding what I should look into?
